### ISSUE TITLE
[Doc] Official code from FrontierLabs for the paper "F5R-TTS: Improving Flow Matching based Text-to-Speech with Group Relative Policy Optimization"

### ISSUE THREAD

**Issue by SaibaLucky** [SWivid/F5-TTS#957]
### Checks

- [x] This template is only for research question, not usage problems, feature requests or bug reports.
- [x] I have thoroughly reviewed the project documentation and read the related paper(s).
- [x] I have searched for existing issues, including closed ones, no similar questions.
- [x] I am using English to submit this issue to facilitate community communication.

### Question details

Hi,

I recently came across the paper "F5R-TTS: Improving Flow Matching based Text-to-Speech with Group Relative Policy Optimization," [https://arxiv.org/pdf/2504.02407](url) which builds upon and improves the existing codebase. The work in the paper seems closely related to this repository.

I was wondering if there are any plans to reproduce the code from this paper?

Looking forward to your response. Thank you!

**Comment by hcsolakoglu** [comment]
I used ai to take a quick look at the repo and paper. Phase 1 and Phase 2 are straightforward. Phase 3 looks doable, but it will take some time. I don't have capacity for this right now. I'll see what I can manage when I get some free time.

**Comment by fakerybakery** [comment]
Would also be interested in seeing an open reproduction of the paper

**Comment by anonymousme** [comment]
I'm from the research group of F5R-TTS, and we have already planned to open-source the code, which will be released soon. Thank you for your attention. 

@SaibaLucky @hcsolakoglu @fakerybakery

**Comment by fakerybakery** [comment]
Hi @anonymousme, is there any update on the release of F5R-TTS?

**Comment by liuxin99** [comment]
<img width="869" alt="Image" src="https://github.com/user-attachments/assets/5fa3fcb6-be84-408d-87a2-4e079e249bb4" />
hi, @anonymousme , Could you please explain how the sample-level sampling probability πθ as shown in the formula is calculated?

**Comment by anonymousme** [comment]
> Hi [@anonymousme](https://github.com/anonymousme), is there any update on the release of F5R-TTS?

@fakerybakery We are organizing and standardizing the codes for better understanding and usage.  I will tag you once the code is released~

**Comment by anonymousme** [comment]
> <img alt="Image" width="869" src="https://private-user-images.githubusercontent.com/33060143/444369246-5fa3fcb6-be84-408d-87a2-4e079e249bb4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3MzM0NDcsIm5iZiI6MTc0NzczMzE0NywicGF0aCI6Ii8zMzA2MDE0My80NDQzNjkyNDYtNWZhM2ZjYjYtYmU4NC00MDhkLTg3YTItNGUwNzllMjQ5YmI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIwVDA5MjU0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwYTgxNDBmZWNjNWUyZjhiZjgxMjE2MmEwNGJkOTRkNmFmMTg2OTFjZDQzZjg0ZmIwM2I5ZjUyZTc2MDI1NzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YoFdYiOml6EvYY66wxDK_U9j33eGB5Rl1lp38nM0eys"> hi, [@anonymousme](https://github.com/anonymousme) , Could you please explain how the sample-level sampling probability πθ as shown in the formula is calculated?

As mentioned in the paper, the output of the model prediction consists of the mean μ and variance σ² of a Gaussian distribution. By sampling a value x from this Gaussian distribution, the corresponding probability density πθ of x can be described by the following formula.

<img width="259" alt="Image" src="https://github.com/user-attachments/assets/d6d2e69e-cea3-4689-9ced-9e302a89cdc7" />

**Comment by anonymousme** [comment]
> > > <img alt="Image" width="869" src="https://private-user-images.githubusercontent.com/33060143/444369246-5fa3fcb6-be84-408d-87a2-4e079e249bb4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3MzM0NDcsIm5iZiI6MTc0NzczMzE0NywicGF0aCI6Ii8zMzA2MDE0My80NDQzNjkyNDYtNWZhM2ZjYjYtYmU4NC00MDhkLTg3YTItNGUwNzllMjQ5YmI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIwVDA5MjU0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwYTgxNDBmZWNjNWUyZjhiZjgxMjE2MmEwNGJkOTRkNmFmMTg2OTFjZDQzZjg0ZmIwM2I5ZjUyZTc2MDI1NzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YoFdYiOml6EvYY66wxDK_U9j33eGB5Rl1lp38nM0eys"> hi, [@anonymousme](https://github.com/anonymousme) , Could you please explain how the sample-level sampling probability πθ as shown in the formula is calculated?
> > 
> > 
> > As mentioned in the paper, the output of the model prediction consists of the mean μ and variance σ² of a Gaussian distribution. By sampling a value x from this Gaussian distribution, the corresponding probability density πθ of x can be described by the following formula.
> > <img alt="Image" width="259" src="https://private-user-images.githubusercontent.com/44261915/445533298-d6d2e69e-cea3-4689-9ced-9e302a89cdc7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc4MjI4MzMsIm5iZiI6MTc0NzgyMjUzMywicGF0aCI6Ii80NDI2MTkxNS80NDU1MzMyOTgtZDZkMmU2OWUtY2VhMy00Njg5LTljZWQtOWUzMDJhODljZGM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIxVDEwMTUzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyYmJiM2E0MTY0ZmUxZjgzNTMzMmQ0M2NkNjhiYzdmNzIyNzZmY2UwYTVkODlhNDhlNTQxMTA2ZTkyMGI3NTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2w1gmjBGaFcTObcoZnA-_i8WPihRC6_1r7Jrf3NQiP4">
> 
> [@anonymousme](https://github.com/anonymousme) Thank you very much for your reply!
> 
> So, judging from the loss function, are you implementing on-policy RL training?
> 
> Additionally, I have another question: You mentioned that 100 hours of data were sampled from the training set for GRPO training. However, when calculating the Speaker Similarity reward, shouldn't the input consist of a prompt audio and a prompt text? How do you organize the training data for this specific scenario? Specifically, does the training data include pairs of utterances from the same speaker but with different content?
> 
> <img alt="Image" width="1029" src="https://private-user-images.githubusercontent.com/33060143/446028839-27e64c93-5e1e-4a47-8023-0561f4ddd6b4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc5ODk4NTMsIm5iZiI6MTc0Nzk4OTU1MywicGF0aCI6Ii8zMzA2MDE0My80NDYwMjg4MzktMjdlNjRjOTMtNWUxZS00YTQ3LTgwMjMtMDU2MWY0ZGRkNmI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIzVDA4MzkxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlZDI5ZDkwNjJhY2NiY2Q5ZmYwZGU0NDRiNzUxZGVjMzUxY2Q3MGZhNzIxMTFlMGMwOGUwMzU5NjE1ODJhODgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-V2sY8HHUoj1ZrpcgzVNau1zhsLEk1m-9fPSwVCouoU"> It seems that there is no prompt in the input condition...🤔

1. Yes, we are implementing on-policy RL training. 
2. During training, a pair of audio and text is given. ​x_gt​ represents the audio, and ​text_gt​ is the corresponding text for this audio. We randomly select a small segment from the ​beginning​ of ​x_gt​ as the prompt audio and mask the remaining part. The model then predicts the masked part, and speaker similarity is calculated between the predicted audio and the ground truth audio.
3. There are some utterances from the same speaker but with different content.

**Comment by yiwei0730** [comment]
@anonymousme  When is the release date of F5R-TTS expected? Thank you!

**Comment by anonymousme** [comment]
The code of F5R-TTS is already released at [https://github.com/FrontierLabs/F5R-TTS](https://github.com/FrontierLabs/F5R-TTS). Thank you for your attention. 
@SaibaLucky @hcsolakoglu @fakerybakery @liuxin99 @yiwei0730

**Comment by QinhanJ** [comment]
I have a dumb question, if this changed training method, will the f5-tts able to use f5r-tts model to infer?
Or the infer logic changed in f5r-tts too?
